javaslang/src/main/java/javaslang/control/Option.java

```diff
@@ -54,14 +59,14 @@
      */
     static <T> Option<Seq<T>> sequence(Iterable<? extends Option<? extends T>> values) {
         Objects.requireNonNull(values, "values is null");
-        List<T> list = List.empty();
+        Stream<T> stream = Stream.empty();
         for (Option<? extends T> value : values) {
             if (value.isEmpty()) {
                 return Option.none();
             }
-            list = list.prepend(value.get());
+            stream = stream.append(value.get());
         }
-        return Option.some(list.reverse());
+        return Option.some(stream);
```

Comment:
@ashrko619 The sequence operation isn't lazy by nature. We have to evaluate all contained elements to decide whether the result is empty or not. Therefore we should choose an eager Seq implementation. Namely either we stay with `List` (because of good memory footprint) or choose `Vector` (because we safe the reverse operation).

I tend to use `Vector`.

---

@paplorinc It is the 'algebraic' definition of `sequence`:

```java
sequence(Seq[M[T]]) =::= M[Seq[T]]
```

The `sequence` operation on a Value (resp. Monad) reflects the properties of a Functor (first failing operation =&gt; empty result) in contrast to an Applicative (subsequent results are collected).